cpp/fetcher/remote_peer.cc

```diff
@@ -67,28 +80,62 @@ void RemotePeer::Impl::DoneGetSTH(
     return;
   }
 
-  // TODO(pphaneuf): At least some of the CHECK-failing here might
-  // have to be changed to... something else.
-  CHECK_EQ(verifier_->VerifySignedTreeHead(
-               *new_sth, 0, (time(NULL) + 10) * kNumMillisPerSecond),
-           LogVerifier::VERIFY_OK)
-      << "could not validate STH:\n" << new_sth->DebugString();
-
-  // TODO(alcutter): Need to check STH consistency here, including for older
-  // STHs which might come in.
-
-  lock_guard<mutex> lock(lock_);
-
-  if (new_sth->tree_size() < 0) {
-    LOG(WARNING) << "Unexpected tree size in new_sth:\n"
-                 << new_sth->DebugString();
-  } else if (sth_ && new_sth->tree_size() < sth_->tree_size()) {
-    LOG(WARNING) << "Received old STH:\n" << new_sth->DebugString();
-  } else if (!sth_ || new_sth->timestamp() > sth_->timestamp()) {
-    // This STH is good, we'll take it.
-    sth_ = new_sth;
-    if (on_new_sth_) {
-      on_new_sth_(*sth_);
+  bool sth_provisionally_valid(true);
+
+  const LogVerifier::VerifyResult result(verifier_->VerifySignedTreeHead(
+      *new_sth, 0, (time(NULL) + 10) * kNumMillisPerSecond));
+
+  // TODO(alcutter): We should probably log these invalid STHs somewhere a bit
+  // more durable.
+  switch (result) {
+    case LogVerifier::VERIFY_OK:
+      // Alright!
+      break;
+    case LogVerifier::INVALID_TIMESTAMP:
+      LOG(WARNING) << "Invalid timestamp on received STH:\n"
+                   << new_sth->DebugString();
+      invalid_sths_received->Increment("invalid_timestamp");
+      sth_provisionally_valid = false;
+      break;
+    case LogVerifier::INVALID_SIGNATURE:
+      LOG(WARNING) << "Invalid signature on received STH:\n"
+                   << new_sth->DebugString();
+      invalid_sths_received->Increment("invalid_signature");
+      sth_provisionally_valid = false;
+      break;
+    case LogVerifier::INCONSISTENT_TIMESTAMPS:
+    case LogVerifier::INVALID_MERKLE_PATH:
+    case LogVerifier::INVALID_FORMAT:
+      sth_provisionally_valid = false;
+      LOG(FATAL) << "Unexpected verify result for VerifySignedTreeHead(): "
+                 << result;
+  }
+
+  const string local_root_at_snapshot(
+      log_lookup_->RootAtSnapshot(new_sth->tree_size()));
+  if (new_sth->sha256_root_hash() != local_root_at_snapshot) {
```

Comment:
I've already LGTM'd this, but I came back to check that #761 indeed had everything I wanted...

This implies that we already have the entries for the STH that we received? Isn't the interesting case the one where you have N entries, get a new STH with N+X entries, triggering the fetching of those next X entries?